File: WebSocketUnityClient/Assets/WebSocketClient/Runtime/Scripts/Logic/WSBackend.cs
using Newtonsoft.Json.Linq;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Threading.Tasks;
using UIFramework;
using UnityEngine;

namespace WebSocketClient
{
    /// <summary>
    /// 负责与服务器进行通讯管理，通讯由一个 wsclient Proxy 实际负责收发
    /// </summary>
    public class WSBackend
    {
        public enum WSBackendState
        {
            Close,
            Open
        }

        public WSBackendState State { get; private set; } = WSBackendState.Close;

        public event Action OnBackendStateChanged;

        public static WSBackend singleton { get; private set; }

        public MonoBehaviour monoGameObject { get; private set; }

        private WSBackend()
        { }

        public static bool CreateSingleton(MonoBehaviour monoBehaviour)
        {
            if (singleton == null)
            {
                singleton = new WSBackend();
                singleton.monoGameObject = monoBehaviour;
                return true;
            }
            return false;
        }

        private bool _isInited = false;

        private string _clientId;

        /// <summary>
        /// 实际负责收发的通讯代理
        /// </summary>
        private WebSocketClient _wsClient;

        public void Init(string url)
        {
            if (!_isInited)
            {
                _clientId = GetClientConnectionId();
                _wsClient = new WebSocketClient(url + $"?clientId={_clientId}");
                monoGameObject.StartCoroutine(HandleReceivedMsg());
                _wsClient.OnClientStateChanged += OnClientStateChanged;           
                _isInited = true;
            }
            else
            {
                Utility.LogDebug("Backend", "Already inited");
            }
        }

        private void OnClientStateChanged()
        {
            if (_wsClient.State == WebSocketClient.ClientState.open)
            {
                if (State != WSBackendState.Open)
                {
                    State = WSBackendState.Open;
                    OnBackendStateChanged?.Invoke();
                }
            }
            else
            {
                if (State != WSBackendState.Close)
                {
                    State = WSBackendState.Close;
                    OnBackendStateChanged?.Invoke();
                }
            }
        }

        /// <summary>
        /// 开始连接服务器
        /// </summary>
        /// <returns></returns>
        public async Task ConnectAndRecvAsync()
        {
            while (Application.isPlaying)
            {
                // 如果连接成功，会阻塞线程
                await _wsClient.ConnectAndRecvAsync();
                Utility.LogDebug("Backend", "connect to ws server fail, reconnect start after 1 sec", _wsClient.State);
                await Task.Delay(1000);
            }

            Utility.LogDebug("Backend", "Reconnect stop by app exit", _wsClient.State);
        }
       
        private IEnumerator HandleReceivedMsg()
        {         
            while (true)
            {
                foreach (var receivedMessage in _wsClient.GetCurrentRecievedDataFromQueue())
                {
                    var data = JObject.Parse(receivedMessage);
                    if (data == null)
                    {
                        continue;
                    }
                    var pack = MsgPack.Parse(data);
                    if (pack != null)
                    {
                        switch (pack.type)
                        {
                            case MsgPack.RequestType:
                                if (pack is RequestPack request)
                                {
                                    OnBackendRequest?.Invoke(request);
                                }
                                break;
                            case MsgPack.ResponseType:
                                if (pack is ResponsePack response)
                                {
                                    OnBackendResponse?.Invoke(response);
                                }
                                break;
                            case MsgPack.NotifyType:
                                if (pack is NotifyPack notify)
                                {
                                    OnBackendNotify?.Invoke(notify);
                                }
                                break;
                            default:
                                break;
                        }
                    }
                    else
                    {
                        Utility.LogInternalDebug($"[WebSocket Client] Received message from server: {receivedMessage} Deserialize fail");
                    }
                }
                yield return new WaitForEndOfFrame();
            }
        }

        /// <summary>
        /// 收到 backend 请求消息 必须回执
        /// </summary>
        public event Action<RequestPack> OnBackendRequest;

        /// <summary>
        /// 收到 backend 回执消息
        /// </summary>
        public event Action<ResponsePack> OnBackendResponse;

        /// <summary>
        /// 收到 backend Notify
        /// </summary>
        public event Action<NotifyPack> OnBackendNotify;

        public class BackendRequestResult
        {
            public WebSocketClient.SendMsgState state;
            public int rid;
        }

        /// <summary>
        /// 向 backend 发送请求
        /// </summary>
        /// <param name="serviceName"></param>
        /// <param name="cmd"></param>
        /// <param name="data"></param>
        /// <returns> rid (request id) </returns>
        public async Task<BackendRequestResult> CreateBackendRequest(string serviceName, string cmd, JToken data)
        {
            if (_wsClient != null && _wsClient.State == WebSocketClient.ClientState.open)
            {
                RequestPack requestPack = new RequestPack()
                {
                    rid = RequestPack.GetRequestId(),
                    clientId = _clientId,
                    cmd = cmd,
                    data = data,
                    serviceName = serviceName
                };
                var sendMsg = requestPack.ToString();
                var state = await _wsClient.SendMessageAsync(sendMsg);
                return new BackendRequestResult()
                {
                    state = state,
                    rid = requestPack.rid
                };
            }
            return new BackendRequestResult()
            {
                state = WebSocketClient.SendMsgState.Unkown,
                rid = -1
            };
        }

        public async Task CloseAsync()
        {
             await _wsClient?.CloseAsync();
        }

        /// <summary>
        /// 获取 clientId
        /// </summary>
        /// <param name="extroInfo"></param>
        /// <returns></returns>
        private static string GetClientConnectionId()
        {
            return SystemInfo.deviceType.ToString() + "_" + Utility.GetDeviceId();
        }
    }
}

File: WebsocketServer/WebsocketServer/wwwroot/TypeScriptClient/lib/clientKernal/WSBackend.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var WebsocketTSClient;
(function (WebsocketTSClient) {
    /**
     * 封装一层 backend
     */
    class WSBackend {
        constructor(_backendUrl) {
            this._backendUrl = _backendUrl;
            this._isInited = false;
            /** ws 连接状态变化 */
            this.OnStateChanged = new WebsocketTSClient.EventHandler();
            /** ws 收到 Response */
            this.OnResponse = new WebsocketTSClient.EventHandler();
            /** ws 收到 Notify */
            this.OnNotify = new WebsocketTSClient.EventHandler();
            this.OnWSClientStateChanged = (state) => {
                this.OnStateChanged.Trigger(state);
            };
            this.OnWSClientMsgRecieved = (msg) => {
                var jobj = JSON.parse(msg);
                if (jobj != null) {
                    if (jobj.type == "response") {
                        if (this.OnResponse != null) {
                            // 收到response
                            this.OnResponse.Trigger(jobj);
                        }
                    }
                    else if (jobj.type == "notify") {
                        if (this.OnNotify != null) {
                            // 收到notify
                            this.OnNotify.Trigger(jobj);
                        }
                    }
                }
            };
            this._uid = WebsocketTSClient.Utility.GenerateUniqueId();
            this._wsClient = new WebsocketTSClient.WebSocketClient(this._backendUrl, this._uid);
            this._wsClient.OnStateChanged.AddListener(this.OnWSClientStateChanged);
            this._wsClient.OnMessageReceived.AddListener(this.OnWSClientMsgRecieved);
        }
        static CreateSingleton(backendUrl) {
            if (!WSBackend.singleton) {
                WSBackend.singleton = new WSBackend(backendUrl);
                return true;
            }
            return false;
        }
        IsConnected() {
            if (this._wsClient == null) {
                return false;
            }
            return this._wsClient.IsConnected();
        }
        Connect2Server() {
            return __awaiter(this, void 0, void 0, function* () {
                if (!this._isInited) {
                    this._wsClient.Connect();
                    const promise = new Promise((resolve, reject) => {
                        const handler = (state) => {
                            if (state) {
                                WebsocketTSClient.Utility.LogDebug("[WSBackend]", "Connect2Server Promise resolved");
                                this._wsClient.OnStateChanged.RmListener(handler);
                                resolve();
                            }
                            // reject 不会使用，因为client会自动断线重连
                        };
                        this._wsClient.OnStateChanged.AddListener(handler);
                    });
                    this._isInited = true;
                    return promise; // Return the promise to be awaited
                }
                else {
                    WebsocketTSClient.Utility.LogDebug("[WSBackend]", "Already inited");
                    return Promise.resolve(); // Return a resolved promise if already initialized
                }
            });
        }
        /**
         * 创建一个请求
         * @param serviceName
         * @param cmd
         * @param data
         * @returns
         */
        CreateBackendRequest(serviceName, cmd, data) {
            if (this._wsClient == null || !this._wsClient.IsConnected) {
                return -1;
            }
            var pack = new WebsocketTSClient.RequestPack(this._uid);
            pack.serviceName = serviceName;
            pack.cmd = cmd;
            pack.data = data;
            this._wsClient.SendMsg(JSON.stringify(pack).replace(/[\r\n\s]/g, ""));
            return pack.rid;
        }
    }
    WebsocketTSClient.WSBackend = WSBackend;
})(WebsocketTSClient || (WebsocketTSClient = {}));
